File: ios/Simple MP3/Data/M3uPlaylist.swift
//
//  M3uPlaylist.swift
//  Simple MP3
//

import Foundation

enum M3uPlaylist {
    struct Entry: Equatable, Sendable {
        var location: String
        var title: String?
        var artist: String?
        var durationSec: Int?

        var fileName: String {
            let trimmed = location.split(separator: "?").first.map(String.init) ?? location
            return (trimmed as NSString).lastPathComponent
        }
    }

    struct MatchResult: Sendable {
        var matched: [Track]
        var unmatched: [Entry]
        var playlistName: String
    }

    static func parse(_ text: String, defaultName: String = "Imported playlist") -> (String, [Entry]) {
        let lines = text.replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
            .components(separatedBy: "\n")
        var entries: [Entry] = []
        var pendingTitle: String?
        var pendingArtist: String?
        var pendingDuration: Int?
        var playlistName = defaultName

        for raw in lines {
            let line = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if line.isEmpty { continue }
            if line.caseInsensitiveCompare("#EXTM3U") == .orderedSame { continue }
            if line.uppercased().hasPrefix("#PLAYLIST:") {
                let name = String(line.dropFirst("#PLAYLIST:".count)).trimmingCharacters(in: .whitespaces)
                if !name.isEmpty { playlistName = name }
                continue
            }
            if line.uppercased().hasPrefix("#EXTINF:") {
                let body = String(line.dropFirst("#EXTINF:".count))
                let comma = body.firstIndex(of: ",")
                let durationPart = comma.map { String(body[..<$0]) } ?? body
                let titlePart = comma.map { String(body[body.index(after: $0)...]).trimmingCharacters(in: .whitespaces) } ?? ""
                pendingDuration = Int(durationPart.trimmingCharacters(in: .whitespaces).split(separator: ".").first.map(String.init) ?? "")
                let parsed = splitArtistTitle(titlePart)
                pendingArtist = parsed.0
                pendingTitle = parsed.1
                continue
            }
            if line.hasPrefix("#") { continue }
            entries.append(Entry(location: line, title: pendingTitle, artist: pendingArtist, durationSec: pendingDuration))
            pendingTitle = nil
            pendingArtist = nil
            pendingDuration = nil
        }
        return (playlistName, entries)
    }

    static func write(name: String, tracks: [Track]) -> String {
        var sb = "#EXTM3U\n"
        if !name.trimmingCharacters(in: .whitespaces).isEmpty {
            sb += "#PLAYLIST:\(name.trimmingCharacters(in: .whitespacesAndNewlines))\n"
        }
        for track in tracks {
            let seconds = track.duration > 0 ? Int(track.duration / 1000) : -1
            sb += "#EXTINF:\(seconds),\(track.artist) - \(track.title)\n"
            sb += "\(track.uri)\n"
        }
        return sb
    }

    static func match(entries: [Entry], library: [Track], playlistName: String) -> MatchResult {
        let songs = library.filter { $0.source != .stream && !$0.isHidden }
        var byURI: [String: Track] = [:]
        var byFile: [String: [Track]] = [:]
        for track in songs {
            let loc = normalizeLocation(track.uri)
            if !loc.isEmpty { byURI[loc] = track }
            let name = fileName(of: track.uri).lowercased()
            if !name.isEmpty {
                byFile[name, default: []].append(track)
            }
        }

        var seen = Set<String>()
        var matched: [Track] = []
        var unmatched: [Entry] = []
        for entry in entries {
            if let hit = matchEntry(entry, byURI: byURI, byFile: byFile, songs: songs),
               seen.insert(hit.id).inserted {
                matched.append(hit)
            } else if matchEntry(entry, byURI: byURI, byFile: byFile, songs: songs) == nil {
                unmatched.append(entry)
            }
        }
        return MatchResult(matched: matched, unmatched: unmatched, playlistName: playlistName)
    }

    static func splitArtistTitle(_ raw: String) -> (String?, String?) {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.isEmpty { return (nil, nil) }
        if let range = text.range(of: " - ") {
            let artist = String(text[..<range.lowerBound]).trimmingCharacters(in: .whitespaces)
            let title = String(text[range.upperBound...]).trimmingCharacters(in: .whitespaces)
            return (artist.isEmpty ? nil : artist, title.isEmpty ? text : title)
        }
        return (nil, text)
    }

    static func normalizeLocation(_ raw: String) -> String {
        var s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.lowercased().hasPrefix("file://") {
            s = String(s.dropFirst("file://".count))
            s = s.removingPercentEncoding ?? s
        }
        return s.replacingOccurrences(of: "\\", with: "/").lowercased()
    }

    static func fileName(of location: String) -> String {
        let trimmed = location.split(separator: "?").first.map(String.init) ?? location
        return (trimmed as NSString).lastPathComponent
    }

    private static func matchEntry(
        _ entry: Entry,
        byURI: [String: Track],
        byFile: [String: [Track]],
        songs: [Track]
    ) -> Track? {
        let loc = normalizeLocation(entry.location)
        if !loc.isEmpty {
            if let hit = byURI[loc] { return hit }
            if let hit = byURI.first(where: { $0.key.hasSuffix(loc) || loc.hasSuffix($0.key) })?.value {
                return hit
            }
        }
        let name = entry.fileName.lowercased()
        if !name.isEmpty, let hits = byFile[name] {
            if hits.count == 1 { return hits[0] }
            if let pick = durationPick(entry, hits) { return pick }
        }
        if let title = entry.title?.trimmingCharacters(in: .whitespaces), !title.isEmpty {
            let hits = songs.filter {
                $0.title.caseInsensitiveCompare(title) == .orderedSame &&
                    (entry.artist == nil || entry.artist?.isEmpty == true
                        || $0.artist.caseInsensitiveCompare(entry.artist!) == .orderedSame)
            }
            if hits.count == 1 { return hits[0] }
            if let pick = durationPick(entry, hits) { return pick }
        }
        return nil
    }

    private static func durationPick(_ entry: Entry, _ candidates: [Track]) -> Track? {
        guard let want = entry.durationSec, want > 0 else { return nil }
        let close = candidates.filter {
            $0.duration > 0 && abs(Int($0.duration / 1000) - want) <= 2
        }
        return close.count == 1 ? close[0] : nil
    }
}

File: android/app/src/main/java/io/karpilabs/simplemp3/data/playlist/M3uPlaylist.kt
package io.karpilabs.simplemp3.data.playlist

import io.karpilabs.simplemp3.data.local.TrackEntity
import java.net.URLDecoder

/**
 * Extended M3U (#EXTM3U) reader/writer. Locations may be content URIs, file paths,
 * or relative names — matching against the local library never auto-imports files.
 */
object M3uPlaylist {
    data class Entry(
        val location: String,
        val title: String? = null,
        val artist: String? = null,
        val durationSec: Int? = null,
    ) {
        val fileName: String
            get() =
                location
                    .substringBefore('?')
                    .substringBefore('#')
                    .trimEnd('/')
                    .substringAfterLast('/')
                    .substringAfterLast('\\')
                    .let { decoded ->
                        runCatching { URLDecoder.decode(decoded, Charsets.UTF_8.name()) }.getOrDefault(decoded)
                    }

        val displayTitle: String
            get() = title?.takeIf { it.isNotBlank() } ?: fileName.substringBeforeLast('.')
    }

    data class MatchResult(
        val matched: List<TrackEntity>,
        val unmatched: List<Entry>,
        val playlistName: String,
    )

    fun parse(
        text: String,
        defaultName: String = "Imported playlist",
    ): Pair<String, List<Entry>> {
        val lines = text.replace("\r\n", "\n").replace('\r', '\n').split('\n')
        val entries = mutableListOf<Entry>()
        var pendingTitle: String? = null
        var pendingArtist: String? = null
        var pendingDuration: Int? = null
        var playlistName = defaultName

        for (raw in lines) {
            val line = raw.trim()
            if (line.isEmpty()) continue
            when {
                line.equals("#EXTM3U", ignoreCase = true) -> Unit
                line.startsWith("#PLAYLIST:", ignoreCase = true) -> {
                    val name = line.substringAfter(':').trim()
                    if (name.isNotEmpty()) playlistName = name
                }
                line.startsWith("#EXTINF:", ignoreCase = true) -> {
                    val body = line.removePrefix("#EXTINF:").removePrefix("#extinf:")
                    val comma = body.indexOf(',')
                    val durationPart = if (comma >= 0) body.substring(0, comma) else body
                    val titlePart = if (comma >= 0) body.substring(comma + 1).trim() else ""
                    pendingDuration = durationPart.trim().substringBefore('.').toIntOrNull()?.takeIf { it >= 0 }
                    val parsed = splitArtistTitle(titlePart)
                    pendingArtist = parsed.first
                    pendingTitle = parsed.second
                }
                line.startsWith("#") -> Unit
                else -> {
                    entries +=
                        Entry(
                            location = line,
                            title = pendingTitle,
                            artist = pendingArtist,
                            durationSec = pendingDuration,
                        )
                    pendingTitle = null
                    pendingArtist = null
                    pendingDuration = null
                }
            }
        }
        return playlistName to entries
    }

    fun write(
        name: String,
        tracks: List<TrackEntity>,
    ): String {
        val sb = StringBuilder()
        sb.append("#EXTM3U\n")
        if (name.isNotBlank()) {
            sb.append("#PLAYLIST:").append(name.trim()).append('\n')
        }
        for (track in tracks) {
            val seconds = if (track.duration > 0) (track.duration / 1000L).toInt() else -1
            val title = "${track.artist} - ${track.title}"
            sb.append("#EXTINF:").append(seconds).append(',').append(title).append('\n')
            sb.append(exportLocation(track)).append('\n')
        }
        return sb.toString()
    }

    fun match(
        entries: List<Entry>,
        library: List<TrackEntity>,
        playlistName: String,
    ): MatchResult {
        if (entries.isEmpty()) {
            return MatchResult(emptyList(), emptyList(), playlistName)
        }
        val songs = library.filter { !it.isStream && !it.isHidden }
        val byUri = HashMap<String, TrackEntity>(songs.size * 2)
        val byFileName = HashMap<String, MutableList<TrackEntity>>()
        for (track in songs) {
            normalizeLocation(track.uri)?.let { byUri[it] = track }
            val name = fileNameOf(track.uri)
            if (name.isNotEmpty()) {
                byFileName.getOrPut(name.lowercase()) { mutableListOf() }.add(track)
            }
        }

        val matched = LinkedHashSet<Long>()
        val matchedTracks = mutableListOf<TrackEntity>()
        val unmatched = mutableListOf<Entry>()

        for (entry in entries) {
            val hit = matchEntry(entry, byUri, byFileName, songs)
            if (hit != null && matched.add(hit.id)) {
                matchedTracks += hit
            } else if (hit == null) {
                unmatched += entry
            }
        }
        return MatchResult(matchedTracks, unmatched, playlistName)
    }

    internal fun splitArtistTitle(raw: String): Pair<String?, String?> {
        val text = raw.trim()
        if (text.isEmpty()) return null to null
        val sep = " - "
        val idx = text.indexOf(sep)
        if (idx <= 0) return null to text
        val artist = text.substring(0, idx).trim().ifBlank { null }
        val title = text.substring(idx + sep.length).trim().ifBlank { text }
        return artist to title
    }

    internal fun normalizeLocation(raw: String): String? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null
        var s = trimmed
        if (s.startsWith("file://", ignoreCase = true)) {
            s = s.removePrefix("file://").removePrefix("FILE://")
            s = runCatching { URLDecoder.decode(s, Charsets.UTF_8.name()) }.getOrDefault(s)
        }
        s = s.replace('\\', '/')
        return s.lowercase()
    }

    internal fun fileNameOf(location: String): String =
        location
            .substringBefore('?')
            .substringBefore('#')
            .trimEnd('/')
            .substringAfterLast('/')
            .substringAfterLast('\\')
            .let { decoded ->
                runCatching { URLDecoder.decode(decoded, Charsets.UTF_8.name()) }.getOrDefault(decoded)
            }

    private fun exportLocation(track: TrackEntity): String = track.uri

    private fun matchEntry(
        entry: Entry,
        byUri: Map<String, TrackEntity>,
        byFileName: Map<String, MutableList<TrackEntity>>,
        songs: List<TrackEntity>,
    ): TrackEntity? {
        normalizeLocation(entry.location)?.let { loc ->
            byUri[loc]?.let { return it }
            // Absolute path may be a suffix of a stored file:// URI or vice versa.
            val suffixHit =
                byUri.entries.firstOrNull { (key, _) ->
                    key.endsWith(loc) || loc.endsWith(key)
                }
            if (suffixHit != null) return suffixHit.value
        }

        val fileName = entry.fileName.lowercase()
        if (fileName.isNotEmpty()) {
            val nameHits = byFileName[fileName]
            if (nameHits != null && nameHits.size == 1) return nameHits[0]
            if (nameHits != null && nameHits.size > 1) {
                durationPick(entry, nameHits)?.let { return it }
            }
        }

        val title = entry.title?.trim().orEmpty()
        if (title.isNotEmpty()) {
            val titleHits =
                songs.filter {
                    it.title.equals(title, ignoreCase = true) &&
                        (
                            entry.artist.isNullOrBlank() ||
                                it.artist.equals(entry.artist, ignoreCase = true)
                        )
                }
            if (titleHits.size == 1) return titleHits[0]
            durationPick(entry, titleHits)?.let { return it }
        }
        return null
    }

    private fun durationPick(
        entry: Entry,
        candidates: List<TrackEntity>,
    ): TrackEntity? {
        val want = entry.durationSec ?: return null
        if (want <= 0) return null
        val close =
            candidates.filter {
                if (it.duration <= 0) return@filter false
                kotlin.math.abs(it.duration / 1000L - want) <= 2
            }
        return close.singleOrNull()
    }
}
